test/Spec/GT/DB/Component.hs

```diff
@@ -0,0 +1,74 @@
+module Spec.GT.DB.Component where
+
+import Spec.GT.DB.Prelude
+
+import Database.Persist.Postgresql
+
+import GT.DB.Schema.Component
+import GT.DB.Schema.Bike
+import qualified GT.DB.Component as Component
+
+spec :: SpecWith TestDb
+spec = do
+    describe "getForBike" $ do
+        itDb "works" $ do
+            let frame = Component
+                    { componentName = "Scrubtrout"
+                    , componentPart = "Frame"
+                    , componentBrand = "Salsa"
+                    , componentModel = "Cutthroat"
+                    , componentDescription = ""
+                    , componentParent = Nothing
+                    }
+            frameId <- insert frame
+            bikeId <- insert $ Bike frameId
+            let wheel = Component
+                    { componentName = "Good wheel"
+                    , componentPart = "Front wheel"
+                    , componentBrand = "WTB"
+                    , componentModel = "Asym 29"
+                    , componentDescription = "it is p good"
+                    , componentParent = Just frameId
+                    }
+            let handlebars = Component
+                    { componentName = "bars"
+                    , componentPart = "handlebars"
+                    , componentBrand = "Salsa"
+                    , componentModel = "Cowchipper 46"
+                    , componentDescription = ""
+                    , componentParent = Just frameId
+                    }
+            wheelId <- insert wheel
+            let tire = Component
+                    { componentName = "knobbss"
+                    , componentPart = "Tire"
+                    , componentBrand = "WTB"
+                    , componentModel = "Riddler 45"
+                    , componentDescription = ""
+                    , componentParent = Just wheelId
+                    }
+            tireId <- insert tire
+            handlebarId <- insert handlebars
+            let frame2 = Component
+                    { componentName = "Chonker"
+                    , componentPart = "Frame"
+                    , componentBrand = "Salsa"
+                    , componentModel = "Mukluk"
+                    , componentDescription = ""
+                    , componentParent = Nothing
+                    }
+            frame2Id <- insert frame2
+            bike2Id <- insert $ Bike frame2Id
+            flatboys <- insert Component
+                { componentName = ""
+                , componentPart = "Handlebars"
+                , componentBrand = "Salsa"
+                , componentModel = "Rustler"
+                , componentDescription = ""
+                , componentParent = Just frame2Id
+                }
+
+            components <- Component.getForBike bikeId
+            map (componentName . entityVal) components
+                `shouldMatchList`
+                    map componentName [ wheel, tire, handlebars, frame ]
```

Comment:
A good test should directly measure what you care about while ignoring stuff you don't care about. Previously, this was ``length components `shouldBe` 2``, which succeeded with the wrong answer! Then the test silently failed because `components` became `Either a [x]`. But then I made this very explicit test, which captured a bug, and made it easy to debug and fix.

Reply:
The scaffolding is all nice, and I'm happy to see it worked out in detail here. I guess the hard part for me, then, is knowing what's worthwhile to test. Can you explain this test a bit, and the reasoning that led you to want to test this function and the reasoning that led you to how to test it this way?

Reply:
E.g. for me it's obvious that if I make an obtuse but fast version of an obvious but slow algorithm, I want to write a test that checks that the two implementations agree (at least on a decent subset of the domain). That's clear. But how do you decide what database tests to write, and how do you decide how to write them?

Reply:
>  Can you explain this test a bit, and the reasoning that led you to want to test this function and the reasoning that led you to how to test it this way?

Yeah, definitely!

The function is defined as a blob of raw SQL because the database library does not support features that are required to express it. So I have a few things I want to verify right off the bat:

1. I wrote syntactically correct SQL (need Postgres to report on this)
2. I wrote SQL that the database library can handle (eg the `??` placeholders need to expand correctly, and I need to ensure I have the relevant columns prepared)

Basically, I need a test that just *runs* the query and doesn't do anything else to prevent somewhat predictable run time errors. I'd say this is testing the 'form' of the code, in the same way that you need GHC to tell you if your Haskell code is well-formed. We need to have `persistent-qq` and `postgres` verify this is well-formed.

After that, there's the *behavior* of the code. I write code and I have some expectation on how it works. I expect this function to return all of the components that belong to a bike, including components that belong to components that belong to the bike, recursively, etc. It's my first recursive CTE ever and I have no idea if I'm doing it right, so testing is a good way to explore this. Playing in the REPL is a good way to explore it, too, but this is automated. I expect the function to also not return any components that are on *other* bikes, or not on the bike, etc. 

TO an experienced DBA, the SQL is probably *obviously* correct. I am not an experienced DBA. But i also want to ensure that *calling* the code is convenient and fine, too! So I want to write a test that captures the things i care about.

This test is Not Ideal because it tests *all* of the things I describe above - if any of them break, the whole test breaks, and it's my job to inspect the test output to figure out *which* property broke. Indeed, a better approach would look like:

```haskell
describe "getForBike" $ before performInserts $ do
    it "is well-formed SQL" (shouldNotThrowException (runQuery getForBike bikeId rideId))
    it "has all components for a bike" (runQuery ... assert has all components)
    it "has no components for another bike" $ ...
    it "has recursive components for a bike" $ ...
```

Anyway, I structure the test to capture all those properties. So first I need the database to have a given state. I need it to contain two bikes, and a bunch of components. Most of those components should belong to the first bike (which I am running the test on), and some of the components should belong to *components* on the first bike (to test recursive part of query).

Then I run the query, and I can make assertions against the output I get. In this case, I just say that "All of the component's returned must be what I inserted."